README.md
# koa-ctrl

<!-- [![NPM version][npm-image]][npm-url] -->
[![License][license-image]][license-url]


> koa-ctrl is used to define controller base koa@v2.

## Install

```sh
$ npm i koa-ctrl --save
```

## API

```js
Controller.prototype.action(name, action)
```

## Example

example-1

```js
const controller = new Controller()

controller.action('doHello', async (ctx, next) => {
  ctx.body = 'hello'
  await next()
})

// response.body -> 'hello'
```

example-2

```js
const controller = new Controller()

controller.action('doHello', {
  before: async (ctx, next) => {
    ctx.body = 'before'
    await next()
  },
  action: async (ctx, next) => {
    ctx.body += '->action'
    await next()
  },
  after: async (ctx, next) => {
    ctx.body += '->after'
    await next()
  }
})

// response.body -> 'before->action->after
```

example-3

```js
const controller = new Controller()

controller.action('doHello', {
  before: async (ctx, next) => {
    ctx.body = 'before'
    await next()
  },
  action: async (ctx, next) => {
    ctx.body += '->action'
    await next()
  }
})

// response.body -> 'before->action
```

## Test

```sh
$ npm test (coverage 100%)
```

## License

MIT


[npm-image]: https://img.shields.io/npm/v/koa-ctrl.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-ctrl
[license-image]: http://img.shields.io/npm/l/koa-compose.svg?style=flat-square
[license-url]: LICENSE


History.md
2.1.0 / 2019-02-26
==================

- rename *koa-controller* to *koa-ctrl*

2.0.0 / 2019-02-26
==================

- Refactor to Supports koa@v2


__tests__/index.js
const request = require('supertest')
const Koa = require('koa')
const Controller = require('..')

describe('parma error', () => {
  it('missing controller name', async () => {
    const controller = new Controller()

    let error
    try {
      controller.action()
    } catch (err) {
      error = err
    }

    expect(error.message).toBe('missing controller name')
  })

  it('missing executable options', async () => {
    const controller = new Controller()

    let error
    try {
      controller.action('doHello')
    } catch (err) {
      error = err
    }

    expect(error.message).toBe('missing executable action')
  })
})

describe('action is function', () => {
  let app, server

  beforeAll(done => {
    app = new Koa()
    server = app.listen(done)
  })

  afterAll(done => {
    server.close(done)
  })

  it('success', async () => {
    const controller = new Controller()

    controller.action('doHello', async (ctx, next) => {
      ctx.body = 'hello'
      await next()
    })

    app.use(controller.doHello)

    await request(server)
      .get('/')
      .expect(200, 'hello')
  })
})

describe('action is object', () => {
  let app, server

  beforeAll(done => {
    app = new Koa()
    server = app.listen(done)
  })

  afterAll(done => {
    server.close(done)
  })

  it('success', async () => {
    const controller = new Controller()

    controller.action('doHello', {
      before: async (ctx, next) => {
        ctx.body = 'before'
        await next()
      },
      action: async (ctx, next) => {
        ctx.body += '->action'
        await next()
      },
      after: async (ctx, next) => {
        ctx.body += '->after'
        await next()
      }
    })

    app.use(controller.doHello)

    await request(server)
      .get('/')
      .expect(200, 'before->action->after')
  })
})


lib/koa-controller.js
/*
 * @Author: xsmallbird
 * @Description:
 * @Date: 2019-02-25 21:26:01
 * @Last Modified by: xsmallbird@gmail.com
 */

const assert = require('assert')
const compose = require('koa-compose')
const EventEmitter = require('events').EventEmitter

// Inherit EventEmitter.prototype
var prototype = Object.create(EventEmitter.prototype)

const LIST = ['before', 'action', 'follow']

prototype.action = function (name, options) {
  let middlewares = []

  assert(name, 'missing name')
  assert(options, 'missing actual action')

  // Accept An Function Param
  if (typeof options === 'function') {
    options = { action: options }
  }

  options = Object.assign({}, {
    before: noop,
    action: noop,
    follow: noop
  }, options) // Options, Sorted

  LIST.forEach(function (item) {
    if (isFunction(options[item])) {
      middlewares.push(options[item])
    }
  })

  let self = this

  this[name] = async function (ctx, next) {
    this.controller = self

    await compose(middlewares)(ctx, next)
  }
}

// Exports module
module.exports = function () {
  return Object.create(prototype)
}

// No operation function
async function noop (ctx, next) {
  await next()
}

// Check value is `Function`
function isFunction (value) {
  return typeof value === 'function'
}


lib/Controller.js
/*
 * @Author: xsmallbird
 * @Description:
 * @Date: 2019-02-25 21:26:01
 * @Last Modified by: xsmallbird@gmail.com
 */

const assert = require('assert')
const compose = require('koa-compose')
const EventEmitter = require('events').EventEmitter

// Inherit EventEmitter.prototype
var prototype = Object.create(EventEmitter.prototype)

const LIST = ['before', 'action', 'after']

prototype.action = function (name, action) {
  let middlewares = []

  assert(name, 'missing controller name')
  assert(action, 'missing executable action')

  // Accept An Function Param
  if (typeof action === 'function') {
    action = { action: action }
  }

  action = Object.assign({}, {
    before: noop,
    action: noop,
    after: noop
  }, action) // option, Sorted

  LIST.forEach(function (item) {
    if (isFunction(action[item])) {
      middlewares.push(action[item])
    }
  })

  let self = this

  this[name] = async function (ctx, next) {
    this.controller = self

    await compose(middlewares)(ctx, next)
  }
}

// Exports module
module.exports = function () {
  return Object.create(prototype)
}

// No operation function
async function noop (ctx, next) {
  await next()
}

// Check value is `Function`
function isFunction (value) {
  return typeof value === 'function'
}
